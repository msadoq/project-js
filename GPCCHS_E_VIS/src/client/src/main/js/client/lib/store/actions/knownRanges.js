import simple from '../helpers/simpleActionCreator';
import * as types from '../types';

// intervals can be a simple array for 1 interval or an array of arrays for multiple intervals
export const addKnownRanges = simple(types.WS_KNOWNINTERVAL_ADD, 'tbdId', 'intervals');
export const removeKnownRanges = simple(types.WS_KNOWNINTERVAL_DELETE, 'tbdId', 'intervals');
