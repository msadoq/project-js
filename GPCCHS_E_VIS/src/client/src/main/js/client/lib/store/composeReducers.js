import __ from 'lodash/fp';

const injectAction = action => reducer => state => reducer(state, action);

/* ---------------------------- composeReducers ----------------------------- */
// Purpose :           take some reducers and return a single reducers.
// example :           composeReducers(r1, r2, r3)
// is equivalent to :  (state, action) => r1(r2(r3(state, action), action), action)

export default (...reducers) => (state, action) => {
  const reducersWithAction = __.map(injectAction(action), reducers);
  return __.compose(...reducersWithAction)(state, action);
};
