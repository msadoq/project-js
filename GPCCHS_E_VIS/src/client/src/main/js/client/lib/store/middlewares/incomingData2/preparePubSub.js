import * as types from '../../types';

const preparePubSub = () => ({ dispatch }) => next => (action) => { // eslint-disable-line
  if (action.type === types.INCOMING_PUBSUB_DATA) {
    console.log('[PreparePubSubMiddleware] ON_INCOMING_PUBSUB_DATA action');
  }
  return next(action);
};

export default preparePubSub;
