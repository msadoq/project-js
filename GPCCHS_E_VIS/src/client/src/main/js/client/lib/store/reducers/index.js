// ====================================================================
// HISTORY
// VERSION : 1.1.0 : : : 28/02/2017 : Initial version
// VERSION : 1.1.2 : DM : #3622 : 10/02/2017 : Move views reducer at top-level in reducers folder
// VERSION : 1.1.2 : DM : #3622 : 14/02/2017 : Show / Hide Text Editor Window
// VERSION : 1.1.2 : DM : #3622 : 17/02/2017 : Merge branch 'dev' into abesson-html-editor
// VERSION : 1.1.2 : DM : #3622 : 17/02/2017 : Creation of timebarTimelines reducer .
// VERSION : 1.1.2 : DM : #3622 : 07/03/2017 : first draft on inspector: retrieve data from rtd on right-click
// VERSION : 1.1.2 : DM : #3622 : 13/03/2017 : Simplify rootReducer, transforming closePage in a thunk
// VERSION : 1.1.2 : DM : #3622 : 13/03/2017 : Fix unremoved views when close a page
// VERSION : 1.1.2 : DM : #5828 : 16/03/2017 : Write first configurationReducer (TextView) .
// VERSION : 1.1.2 : DM : #5822 : 20/03/2017 : merge dev in working branch
// VERSION : 1.1.2 : DM : #5828 : 20/03/2017 : Creation of data store for plotView
// VERSION : 1.1.2 : DM : #5822 : 21/03/2017 : merge dev in working branch
// VERSION : 1.1.2 : DM : #3622 : 22/03/2017 : Update viewData organization for last structure + cleaning
// VERSION : 1.1.2 : DM : #5822 : 23/03/2017 : merge dev in working branch
// VERSION : 1.1.2 : DM : #5828 : 07/04/2017 : add entry points to mimic view
// VERSION : 1.1.2 : DM : #5828 : 11/04/2017 : Implement automation for data reducers in viewManager
// VERSION : 1.1.2 : DM : #5828 : 12/04/2017 : New GenericModal component displayed or not displayed at root (Window.js) AddTimeline and EditTimeline forms displayed through it.
// VERSION : 1.1.2 : DM : #6129 : 04/05/2017 : merge dev on mimic branch
// VERSION : 1.1.2 : DM : #5828 : 09/05/2017 : Plot & Text editor panels and sub-panels are stored in store.
// VERSION : 1.1.2 : DM : #5828 : 10/05/2017 : Plot & Text editor panels and sub-panels are stored in store.
// VERSION : 1.1.2 : DM : #6129 : 19/06/2017 : moved components/animations in separate files. Possibility to add it in editor using context menu
// VERSION : 1.1.2 : DM : #6688 : 05/07/2017 : First draft on catalog explorer
// VERSION : 1.1.2 : DM : #6688 : 05/07/2017 : catalog explorer : open, close and browse items
// VERSION : 1.1.2 : DM : #6700 : 24/07/2017 : CReation of knownRanges reducer and actions
// END-HISTORY
// ====================================================================

import { combineReducers } from 'redux';
import { reducer as form } from 'redux-form';

import sessions from './sessions';
import timebars from './timebars';
import views from './views';
import messages from './messages';
import modals from './modals';
import timelines from './timelines';
import windows from './windows';
import pages from './pages';
import domains from './domains';
import timebarTimelines from './timebarTimelines';
import masterSession from './masterSession';
import health from './health';
import codeEditor from './codeEditor';
import inspector from './inspector';
import knownRanges from './knownRanges';
import ui from './ui';
import catalogs from './catalogs';
import rte from './rte';
import comObjectMap from './comObjectMap';
import cache from './cache';
import { getConfigurationReducers, getDataReducers, getUiReducers } from '../../viewManager/reducers';
import hsc from './hsc';

const rootReducer = combineReducers({
  cache,
  form,
  hsc,
  timebars,
  timebarTimelines,
  messages,
  modals,
  timelines,
  windows,
  pages,
  views,
  domains,
  sessions,
  masterSession,
  health,
  codeEditor,
  inspector,
  ui,
  catalogs,
  rte,
  knownRanges,
  comObjectMap,
  ...getConfigurationReducers(),
  ...getDataReducers(),
  ...getUiReducers(),
});

export default rootReducer;
