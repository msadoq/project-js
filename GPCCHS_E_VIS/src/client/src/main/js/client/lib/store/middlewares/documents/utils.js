import _ from 'lodash/fp';
import parameters from 'common/configurationManager';
import { getWorkspaceFolder } from 'store/reducers/hsc';
import { EXTENSIONS } from 'constants';

export const isView = docType => /^.*View$/.test(docType);

const getFilter = docType => (
  { name: docType, extensions: [EXTENSIONS[docType]] }
);

const allViewsFilters = _.keys(EXTENSIONS).filter(isView).map(getFilter);
const allViewsExtensions = _.flatMap(_.get('extensions'), allViewsFilters);

export const getOpenExtensionsFilters = (docType) => {
  if (docType && !isView(docType)) {
    return [getFilter(docType)];
  }
  return [
    { name: 'Any VIMA views', extensions: allViewsExtensions },
    ...allViewsFilters,
  ];
};

export const getSaveExtensionsFilters = docType => [getFilter(docType)];

// warning, this selector is impure, do not memoize it
export const getDefaultFolder = state => (
  getWorkspaceFolder(state) || parameters.get('ISIS_DOCUMENTS_ROOT')
);
