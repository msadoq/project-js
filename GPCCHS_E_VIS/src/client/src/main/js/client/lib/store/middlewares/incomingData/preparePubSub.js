// ====================================================================
// HISTORY
// VERSION : 1.1.2 : DM : #6700 : 18/08/2017 : Update multiple test and implementation
// VERSION : 1.1.2 : DM : #6700 : 21/08/2017 : Fix forecast error and fix related tests
// VERSION : 1.1.2 : FA : #7578 : 23/08/2017 : Add throttle mechanism to pubSubController
// VERSION : 1.1.2 : FA : #7578 : 24/08/2017 : Add robustness code on dataId retrieval
// VERSION : 1.1.2 : DM : #6700 : 25/08/2017 : Add redux and patch workflow improvment + remove
//  store observer
// VERSION : 1.1.2 : DM : #6700 : 25/08/2017 : Add execution map trace in three middlewares to make
//  performance analysis easier
// VERSION : 1.1.2 : DM : #6700 : 30/08/2017 : move dumpBuffer use in a specific middleware
// VERSION : 1.1.2 : DM : #6816 : 06/09/2017 : test perfs on hss process
// VERSION : 2.0.0 : DM : #5806 : 13/10/2017 : change middleware behavior to manage alarm pubsub
//  data
// VERSION : 2.0.0 : DM : #5806 : 06/12/2017 : Change all relative imports .
// VERSION : 2.0.0 : FA : ISIS-FT-2159 : 20/03/2018 : editeur champ flowType VIMA JS
// END-HISTORY
// ====================================================================

import _isBuffer from 'lodash/isBuffer';
import { applyFilters } from 'viewManager/commonData/applyFilters';
import {
  isDataIdInCache,
  isTimestampInKnownRanges,
  getKnownRanges,
} from 'store/reducers/knownRanges';
import {
  isDataIdInDatamapLast,
  isTimestampInLastDatamapInterval,
} from 'dataManager/mapSelector';
import { decode, getType } from 'utils/adapters';
import { add as addMessage } from 'store/actions/messages';
import { set as setLastPubSubTimestamp } from 'serverProcess/models/lastPubSubTimestamp';
import executionMonitor from 'common/logManager/execution';
import dataMapGenerator from 'dataManager/map';
import { newData } from 'store/actions/incomingData';
import * as types from 'store/types';
import { PREFIX_KNOWN_RANGES } from 'constants';

const logger = require('../../../common/logManager')('middleware:preparePubSub');

const preparePubSub = lokiManager => ({ dispatch, getState }) => next => (action) => {
  const nextAction = next(action);
  if (action.type !== types.INCOMING_PUBSUB_DATA
    && action.type !== types.INCOMING_PUBSUBALARM_DATA
  ) {
    return nextAction;
  }

  const execution = executionMonitor('middleware:preparePubSub');
  execution.start('global');

  // check payloads parity
  const payloadBuffersMap = action.payload.data;
  const payloadBuffersArray = Object.keys(payloadBuffersMap);
  // Compute dataMap
  execution.start('dataMap');
  const state = getState();
  const dataMap = dataMapGenerator(state);
  execution.stop('dataMap');

  let payloadsJson = {};

  for (let i = 0; i < payloadBuffersArray.length; i += 1) {
    const { dataId, payloadBuffers } = payloadBuffersMap[payloadBuffersArray[i]];
    // get payload type and check validity
    const payloadProtobufType = getType(dataId.comObject);
    if (typeof payloadProtobufType === 'undefined') {
      // TODO dispatch error
      logger.error('unsupported comObject', dataId.comObject);
      execution.stop('global');
      execution.print();
      return nextAction;
    }

    // Retrieves tbdIds corresponding to dataId from store.knownRanges (storeTbdIds)
    execution.start('isDataIdInCache');
    const storeTbdIds = isDataIdInCache(state, { dataId });
    execution.stop('isDataIdInCache');
    // Retrieve tbdIds corresponding to dataId from dataMap.expectedLastInterval (dataMapTbdIds)
    execution.start('isDataIdInDatamapLast');
    const dataMapTbdIds = isDataIdInDatamapLast(state, dataId);
    execution.stop('isDataIdInDatamapLast');

    // Reduce peers as data by decoding timestamp buffers
    // Loop over peers and for each decode timestamp
    let iBuffer = 0;
    while (iBuffer < payloadBuffers.length) {
      // Get elements by peer
      const timeBuffer = payloadBuffers[iBuffer];
      const dataBuffer = payloadBuffers[iBuffer + 1];
      iBuffer += 2;
      try {
        // robustness code, LPISIS could send empty ZeroMQ frame
        if (_isBuffer(timeBuffer) && _isBuffer(dataBuffer)) {
          execution.start('decode timestamp');
          const timestamp = decode('dc.dataControllerUtils.Timestamp', timeBuffer);
          execution.stop('decode timestamp');

          // persist last received pub/sub timestamp
          setLastPubSubTimestamp(timestamp.ms);

          // decode Payload only once by payloadBuffer loop to avoid resource-consuming
          execution.start('decode payload');
          const decodedPayload = decode(payloadProtobufType, dataBuffer);
          execution.stop('decode payload');

          // For each tbdId in storeList
          execution.start('process for ranges');
          for (let j = 0; j < storeTbdIds.length; j += 1) {
            const tbdId = storeTbdIds[j];
            const filters = getKnownRanges(state, { tbdId }).filters;
            payloadsJson = updateFinalPayload(state, {
              tbdId,
              timestamp: timestamp.ms,
              decodedPayload,
              isInIntervals: isTimestampInKnownRanges,
              filters,
              addRecord: lokiManager.addRecord,
              type: PREFIX_KNOWN_RANGES,
            }, payloadsJson);
          }
          execution.stop('process for ranges');

          // For each tbdId in dataMapList
          execution.start('process for lasts');
          for (let k = 0; k < dataMapTbdIds.length; k += 1) {
            const tbdId = dataMapTbdIds[k];
            // If tbdId already present in final object, checks if current timestamp is already stored
            if (!payloadsJson[tbdId] || !payloadsJson[tbdId][timestamp.ms]) {
              const filters = dataMap.perLastTbdId[tbdId].filters;
              payloadsJson = updateFinalPayload(state, {
                tbdId,
                timestamp: timestamp.ms,
                decodedPayload,
                isInIntervals: isTimestampInLastDatamapInterval,
                filters,
              }, payloadsJson);
            }
          }
          execution.stop('process for lasts');
        }
      } catch (e) {
        dispatch(addMessage('global', 'warning', 'error on processing header buffer '.concat(e)));
        logger.error('error on processing buffer', e);
      }
    }
  }
  // dispatch data per tbdId
  const tbdIds = Object.keys(payloadsJson);

  if (tbdIds.length) {
    dispatch(newData(payloadsJson));
  }

  execution.stop('global');
  execution.print();
  return nextAction;
};

/**
 * Add a set of values in a collection for a given tbdId
 * @param state state {string} tbdId: tbdId of data to check
 * @param tbdId {number} timestamp: timestamp to check in ms
 * @param timestamp {Object} decodedPayload: the decoded payload
 * @param decodedPayload {function} isInIntervals: check if data is in intervals (state, { tbdId, timestamp })
 * @param isInIntervals {function} isInInterval
 * @param filters {Array} filters
 * @param addRecord {function} addRecord: function to add record in cache. Can be undefined
 * @param finalPayloads
 * @returns {*} {Object} updated finalPayload
 */
export const updateFinalPayload = (
  state,
  { tbdId, timestamp, decodedPayload, isInIntervals, filters, addRecord, type },
  finalPayloads
) => {
  const updatedPayloads = finalPayloads;
  // If timestamp is in a known interval
  if (isInIntervals(state, { tbdId, timestamp })) {
    // If data validates filters
    if (applyFilters(decodedPayload, filters)) {
      // Store data in cache with the current tbdId
      if (addRecord) {
        addRecord(type, tbdId, { timestamp, payload: decodedPayload });
      }
      // Add data to final object with the current tbdId
      if (!updatedPayloads[tbdId]) {
        updatedPayloads[tbdId] = {};
      }
      updatedPayloads[tbdId][timestamp] = decodedPayload;
    }
  }
  return updatedPayloads;
};

export default preparePubSub;
