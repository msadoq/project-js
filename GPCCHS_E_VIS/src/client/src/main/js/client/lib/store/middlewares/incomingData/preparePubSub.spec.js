// ====================================================================
// HISTORY
// VERSION : 1.1.2 : DM : #6700 : 18/08/2017 : Update multiple test and implementation
// END-HISTORY
// ====================================================================

import configureMockStore from 'redux-mock-store';
import _omit from 'lodash/omit';
import { isTimestampInKnownRanges } from 'store/reducers/knownRanges';
import { isTimestampInLastDatamapInterval } from 'dataManager/mapSelector';
import lokiManager from 'serverProcess/models/lokiKnownRangesData';
import { getStubData } from 'utils/stubs';
import * as types from 'store/types';
import state from 'common/jest/stateTest';
import preparePubSub, { updateFinalPayload } from './preparePubSub';

const { mockRegister, mockLoadStubs } = require('../../../common/jest');

mockRegister();
mockLoadStubs();
const dataStub = getStubData();

describe('store:middlewares:preparePubSub', () => {
  const mockStore = configureMockStore([preparePubSub(lokiManager)]);
  const decodedPayload = {
    extractedValue: { type: 'integer', value: 1 },
    rawValue: { type: 'integer', value: 2 } };

  beforeEach(() => {
    lokiManager.removeAllCollections();
  });
  describe('updateFinalPayload', () => {
    test('in store, tbdId in store, timestamp ok, empty finalPayloads', () => {
      const store = mockStore(state);
      const tbdId = 'Reporting.ATT_BC_REVTCOUNT1<ReportingParameter>:0:1:::';
      const finalPayloads = updateFinalPayload(store.getState(),
        { tbdId,
          timestamp: 320000,
          decodedPayload,
          isInIntervals: isTimestampInKnownRanges,
          filters: [],
          addRecord: lokiManager.addRecord },
        {});
      expect(finalPayloads).toEqual({ [tbdId]: { 320000: decodedPayload } });
      expect(lokiManager.listCachedCollections()).toEqual([tbdId]);
    });
    test('in store, tbdId in store, timestamp nok, empty finalPayloads', () => {
      const store = mockStore(state);
      const finalPayloads = updateFinalPayload(store.getState(),
        { tbdId: 'Reporting.ATT_BC_REVTCOUNT1<ReportingParameter>:0:1:::',
          timestamp: 150,
          decodedPayload,
          isInIntervals: isTimestampInKnownRanges,
          filters: [],
          addRecord: lokiManager.addRecord },
        {});
      expect(finalPayloads).toEqual({ });
      expect(lokiManager.listCachedCollections()).toEqual([]);
      expect(lokiManager.displayCollection('Reporting.ATT_BC_REVTCOUNT1<ReportingParameter>:0:1:::'))
      .toEqual([]);
    });
    test('in store, tbdId not in store, empty finalPayloads', () => {
      const finalPayloads = updateFinalPayload(state,
        { tbdId: 'Reporting.UNKNOWN_NAME<ReportingParameter>:1:1:::',
          timestamp: 15,
          decodedPayload,
          isInIntervals: isTimestampInKnownRanges,
          filters: [],
          addRecord: lokiManager.addRecord },
        {});
      expect(finalPayloads).toEqual({ });
      expect(lokiManager.listCachedCollections()).toEqual([]);
    });
    test('in store, tbdId not in store, finalPayloads not empty', () => {
      const finalPayloads = updateFinalPayload(state,
        { tbdId: 'Reporting.UNKNOWN_NAME<ReportingParameter>:1:1:::',
          timestamp: 15,
          decodedPayload,
          isInIntervals: isTimestampInKnownRanges,
          filters: [],
          addRecord: lokiManager.addRecord },
        { 'Reporting.ATT_BC_REVTCOUNT1<ReportingParameter>:0:1:::': { 15: decodedPayload } });
      expect(finalPayloads)
      .toEqual({ 'Reporting.ATT_BC_REVTCOUNT1<ReportingParameter>:0:1:::': { 15: decodedPayload } });
      expect(lokiManager.listCachedCollections()).toEqual([]);
    });
    test('in store, tbdId in store, timestamp ok, finalPayloads not empty', () => {
      const tbdId1 = 'Reporting.ATT_BC_REVTCOUNT1<ReportingParameter>:0:1:::';
      const tbdId2 = 'Reporting.OTHER_PID<ReportingParameter>:1:1:::';
      const finalPayloads = updateFinalPayload(state,
        { tbdId: tbdId1,
          timestamp: 320000,
          decodedPayload,
          isInIntervals: isTimestampInKnownRanges,
          filters: [],
          addRecord: lokiManager.addRecord },
        { [tbdId1]: { 300000: decodedPayload } });
      expect(finalPayloads)
      .toEqual({ [tbdId1]: { 300000: decodedPayload, 320000: decodedPayload } });
      const finalPayloads2 = updateFinalPayload(state,
        { tbdId: tbdId1,
          timestamp: 320000,
          decodedPayload,
          isInIntervals: isTimestampInKnownRanges,
          filters: [],
          addRecord: lokiManager.addRecord },
        { [tbdId2]: { 300000: decodedPayload } });
      expect(finalPayloads2).toEqual({
        [tbdId1]: { 320000: decodedPayload },
        [tbdId2]: { 300000: decodedPayload },
      });
      expect(lokiManager.listCachedCollections()).toEqual([tbdId1]);
    });
    test('in dataMap, tbdId not in last', () => {
      const finalPayloads = updateFinalPayload(state,
        { tbdId: 'Reporting.UNKNOWN_NAME<ReportingParameter>:1:1:::',
          timestamp: 320000,
          decodedPayload,
          isInIntervals: isTimestampInLastDatamapInterval,
          filters: [] },
        { 'Reporting.ATT_BC_REVTCOUNT1<ReportingParameter>:0:1:::': { 320000: decodedPayload } });
      expect(finalPayloads)
      .toEqual({ 'Reporting.ATT_BC_REVTCOUNT1<ReportingParameter>:0:1:::': { 320000: decodedPayload } });
      expect(lokiManager.listCachedCollections()).toEqual([]);
    });
    test('in dataMap, tbdId in last', () => {
      const finalPayloads = updateFinalPayload(state,
        { tbdId: 'Reporting.AGA_AM_PRIORITY<ReportingParameter>:0:4:::',
          timestamp: 320000,
          decodedPayload,
          isInIntervals: isTimestampInLastDatamapInterval,
          filters: [] },
        { });
      expect(finalPayloads).toEqual({
        'Reporting.AGA_AM_PRIORITY<ReportingParameter>:0:4:::': { 320000: decodedPayload },
      });
      expect(lokiManager.listCachedCollections()).toEqual([]);
    });
    test('in dataMap, tbdId in last but does not validate filters', () => {
      const finalPayloads = updateFinalPayload(state,
        { tbdId: 'Reporting.AGA_AM_PRIORITY<ReportingParameter>:0:4:::',
          timestamp: 320000,
          decodedPayload,
          isInIntervals: isTimestampInLastDatamapInterval,
          filters: [{
            field: 'extractedValue',
            operator: '>',
            operand: '10',
          }] },
        { });
      expect(finalPayloads).toEqual({});
    });
  });
  describe('preparePubSub middleware', () => {
    const t1 = 350000;
    const t2 = 360000;

    const timestampInLast1 = 370000;
    const timestampInLast2 = 380000;

    const timestampNotInLast1 = 600000;
    const timestampNotInLast2 = 610000;

    const timestamp1 = dataStub.getTimestampProtobuf({ ms: t1 });
    const timestamp2 = dataStub.getTimestampProtobuf({ ms: t2 });

    const timestampInLast1Proto = dataStub.getTimestampProtobuf({ ms: timestampInLast1 });
    const timestampInLast2Proto = dataStub.getTimestampProtobuf({ ms: timestampInLast2 });

    const timestampNotInLast1Proto = dataStub.getTimestampProtobuf({ ms: timestampNotInLast1 });
    const timestampNotInLast2Proto = dataStub.getTimestampProtobuf({ ms: timestampNotInLast2 });

    const rp1 = dataStub.getReportingParameter();
    const rp2 = dataStub.getReportingParameter();

    const protoRp1 = dataStub.getReportingParameterProtobuf(rp1);
    const protoRp2 = dataStub.getReportingParameterProtobuf(rp2);

    const dataId = {
      catalog: 'Reporting',
      parameterName: 'TMMGT_BC_VIRTCHAN3',
      comObject: 'ReportingParameter',
      sessionId: 0,
      domainId: 4,
    };
    const incomingRangeData = () => ({
      type: types.INCOMING_PUBSUB_DATA,
      payload: {
        data: {
          'Reporting.TMMGT_BC_VIRTCHAN3<ReportingParameter>:0:4:::': {
            dataId,
            payloadBuffers: [timestamp1, protoRp1, timestamp2, protoRp2],
          },
        },
      },
    });

    const incomingDataNotInLast = () => ({
      type: types.INCOMING_PUBSUB_DATA,
      payload: {
        data: {
          'Reporting.ATT_BC_REVTCOUNT1<ReportingParameter>:0:1:::': {
            dataId,
            payloadBuffers: [
              timestampNotInLast1Proto,
              protoRp1,
              timestampNotInLast2Proto,
              protoRp2,
            ],
          },
        },
      },
    });

    const incomingDataOneInLast = () => ({
      type: types.INCOMING_PUBSUB_DATA,
      payload: {
        data: {
          'Reporting.ATT_BC_REVTCOUNT1<ReportingParameter>:0:1:::': {
            dataId,
            payloadBuffers: [timestampNotInLast1Proto, protoRp1, timestampInLast1Proto, protoRp2],
          },
        },
      },
    });

    const incomingDataAllInLast = () => ({
      type: types.INCOMING_PUBSUB_DATA,
      payload: {
        data: {
          'Reporting.ATT_BC_REVTCOUNT1<ReportingParameter>:0:1:::': {
            dataId,
            payloadBuffers: [timestampInLast1Proto, protoRp1, timestampInLast2Proto, protoRp2],
          },
        },
      },
    });

    const removeDataFromPayload = (actions) => {
      const tabActions = [];
      for (let i = 0; i < actions.length; i += 1) {
        tabActions.push({ ...actions[i],
          payload: _omit(actions[i].payload, ['data', 'peers']),
        });
      }
      return tabActions;
    };

    test('OTHER_ACTION', () => {
      const otherActions = { type: 'WS_OTHER_ACTION' };
      const store = mockStore(state);
      store.dispatch(otherActions);
      expect(store.getActions()).toEqual([{ type: 'WS_OTHER_ACTION' }]);
    });
    test('Range Data to add in store', () => {
      state.pages.page1.views = ['plot1'];
      const store = mockStore(state);
      store.dispatch(incomingRangeData());
      const list = Object.keys(store.getActions()[1].payload.data);
      for (let i = 0; i < list.length; i += 1) {
        expect(Object.keys(store.getActions()[1].payload.data[list[i]]))
      .toEqual([String(t1), String(t2)]);
      }
      expect(removeDataFromPayload(store.getActions())).toMatchSnapshot();
    });
    test('Not in Last', () => {
      state.pages.page1.views = ['text1', 'mimic1'];
      const store = mockStore(state);
      store.dispatch(incomingDataNotInLast());
      expect(store.getActions().length).toBe(1);
    });
    test('One data in Last', () => {
      state.pages.page1.views = ['text1', 'mimic1'];
      const store = mockStore(state);
      store.dispatch(incomingDataOneInLast());
      const list = Object.keys(store.getActions()[1].payload.data);
      for (let i = 0; i < list.length; i += 1) {
        expect(Object.keys(store.getActions()[1].payload.data[list[i]]))
        .toEqual([String(timestampInLast1)]);
      }
      expect(removeDataFromPayload(store.getActions())).toMatchSnapshot();
    });
    test('All data in Last', () => {
      state.pages.page1.views = ['text1', 'mimic1'];
      const store = mockStore(state);
      store.dispatch(incomingDataAllInLast());
      const list = Object.keys(store.getActions()[1].payload.data);
      for (let i = 0; i < list.length; i += 1) {
        expect(Object.keys(store.getActions()[1].payload.data[list[i]]))
        .toEqual([String(timestampInLast1), String(timestampInLast2)]);
      }
      expect(removeDataFromPayload(store.getActions())).toMatchSnapshot();
    });
  });
});
