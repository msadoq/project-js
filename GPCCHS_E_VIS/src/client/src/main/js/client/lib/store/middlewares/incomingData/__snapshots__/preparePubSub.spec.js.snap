// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`store:middlewares:preparePubSub preparePubSub middleware All data in Last 1`] = `
Array [
  Object {
    "payload": Object {},
    "type": "INCOMING_PUBSUB_DATA",
  },
  Object {
    "payload": Object {},
    "type": "NEW_DATA",
  },
]
`;

exports[`store:middlewares:preparePubSub preparePubSub middleware One data in Last 1`] = `
Array [
  Object {
    "payload": Object {},
    "type": "INCOMING_PUBSUB_DATA",
  },
  Object {
    "payload": Object {},
    "type": "NEW_DATA",
  },
]
`;

exports[`store:middlewares:preparePubSub preparePubSub middleware Range Data to add in store 1`] = `
Array [
  Object {
    "payload": Object {},
    "type": "INCOMING_PUBSUB_DATA",
  },
  Object {
    "payload": Object {},
    "type": "NEW_DATA",
  },
]
`;
